<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to MagicHelpR • MagicHelpR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">MagicHelpR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Introduction to MagicHelpR</a>
    </li>
    <li>
      <a href="../articles/retrieving_genes_in_qtl.html">Retrieving genes in QTL</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Introduction to MagicHelpR</h1>
                        <h4 class="author">Hugo Tavares</h4>
            
            <h4 class="date">2018-02-21</h4>
          </div>

    
    
<div class="contents">
<p>This is a package to perform QTL mapping on the <a href="http://mtweb.cs.ucl.ac.uk/mus/www/magic/">Arabidopsis MAGIC lines</a>, described in <a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1000551">Kover P .. Mott R (2009)</a>. It was mostly written for personal convenience, and it follows the same method implemented in the scripts available on that website.</p>
<p>The package allows you to download the genotype data from the public server, and convert it to PLINK format in case that is useful for other analysis. A small collection of functions allows to estimate the genotype probability matrix (using the <code>happy.hbrem</code> package) and run association tests between a phenotype of interest and each marker, using fixed effects linear models. This can be extended with more complex models, by accessing the genotype and phenotype data stored in a structured object created by the package.</p>
<p>This vignette introduces the main functions in the package, using a workflow with example phenotype data downloaded from the <a href="http://mtweb.cs.ucl.ac.uk/mus/www/magic/MAGIC.phenotype.example.12102015.txt">MAGIC website</a>.</p>
<p>We start by loading the package, which also loads the package <code>dplyr</code>, on which it uses several functions from. Through the rest of this vignette, <code>dplyr</code> functions are often used, so it’s best if you’re familiar with their use and the use of “pipes” <code>%&gt;%</code> (see <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/dplyr.html">dplyr introduction</a>)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(MagicHelpR)</code></pre></div>
<pre><code>## Loading required package: dplyr</code></pre>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre><code>## Loading required package: tidyr</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)</code></pre></div>
<pre><code>## ── Attaching packages ────────────────────────────────── tidyverse 1.2.1 ──</code></pre>
<pre><code>## ✔ ggplot2 2.2.1.9000     ✔ purrr   0.2.4     
## ✔ tibble  1.4.2          ✔ stringr 1.2.0     
## ✔ readr   1.1.1          ✔ forcats 0.2.0</code></pre>
<pre><code>## ── Conflicts ───────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<div id="downloading-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#downloading-the-data" class="anchor"></a>Downloading the data</h2>
<p>For convenience, a function is included to download the genotype data. There is an option in the function that also calls the function <code><a href="../reference/tidyArabMagic.html">tidyArabMagic()</a></code>, which “tidies” that data (this mostly converts things to different formats and also combines the separate chromosome files into a single file, for convenience). Another option, which is set to FALSE by default, is to download example phenotype data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/downloadArabMagic.html">downloadArabMagic</a></span>(<span class="st">'~/temp/magic_intro/'</span>, <span class="dt">tidy =</span> <span class="ot">TRUE</span>, <span class="dt">example_data =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="genotype-probability-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#genotype-probability-matrix" class="anchor"></a>Genotype probability matrix</h2>
<p>The idea with the MAGIC lines is to use information about the founder population when doing the association mapping. Therefore, one needs to infer from which of the 19 founder accessions, each marker allele derives from. This “reconstruction” step can be performed using the function <code><a href="../reference/magicFounderReconstruct.html">magicFounderReconstruct()</a></code>, which uses the functions from the <code>happy.hbrem</code> package.</p>
<p>The function requires a directory where all the genotype files are stored. Note that these files should preferably have been produced by the <code><a href="../reference/tidyArabMagic.html">tidyArabMagic()</a></code> function to ensure data formats are correct.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">magic_geno &lt;-<span class="st"> </span><span class="kw"><a href="../reference/magicFounderReconstruct.html">magicFounderReconstruct</a></span>(<span class="dt">snp_dir =</span> <span class="st">"~/temp/magic_intro"</span>)</code></pre></div>
<pre><code>## mindist: 1e-05
## datafile /home/hugo/temp/magic_intro/all_chr.MAGIC.happy.data allelesfile /home/hugo/temp/magic_intro/all_chr.MAGIC.happy.alleles gen 7
## genotype phase: unknown
## Reading phenotype and genotype data from ped file /home/hugo/temp/magic_intro/all_chr.MAGIC.happy.data
## Number of individuals: 703  
## Number of markers:     1259 
## Number of strains:     19   
## Use Parents:           no
## Number of subjects with two parents: 0    
## Number of parents in nuclear families: 0    
## null model mean nan var nan
## assuming haploid(inbred) genotypes
## datafile /home/hugo/temp/magic_intro/all_chr.MAGIC.happy.data allelesfile /home/hugo/temp/magic_intro/all_chr.MAGIC.happy.alleles gen 7</code></pre>
<p>This function returns an object of class “MagicGen”. This object contains the following:</p>
<ul>
<li>a table with all the marker names and their locations. This can be accessed with he function <code><a href="../reference/getMarkers.html">getMarkers()</a></code>.</li>
<li>a list of SNP genotypes for each marker. This can be accessed with the function <code><a href="../reference/getGenotypes.html">getGenotypes()</a></code> adding the option <code>type = "allele"</code>.</li>
<li>a list of genotype probabilities for each marker. This can be access with the function <code><a href="../reference/getGenotypes.html">getGenotypes()</a></code>.</li>
<li>a table with SNP genotypes of the founder accessions.</li>
</ul>
<p>Here are some examples of using these functions to access information from this.</p>
<p>If we want to see marker information</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw"><a href="../reference/getMarkers.html">getMarkers</a></span>(magic_geno))</code></pre></div>
<pre><code>##       marker       map     bp chromosome
## 1  MN1_29291 0.1220500  29291          1
## 2  MN1_29716 0.1239917  29757          1
## 3 MN1_112907 0.4704500 112907          1
## 4  MASC03771 0.7275250 174605          1
## 5 MN1_197787 0.8241167 197787          1
## 6 MN1_340810 1.4200458 340810          1</code></pre>
<p>If we want to see SNP allele genotypes for marker “MN1_29291”:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw"><a href="../reference/getGenotypes.html">getGenotypes</a></span>(magic_geno, <span class="dt">type =</span> <span class="st">"allele"</span>)$MN1_29291)</code></pre></div>
<pre><code>##   MAGIC.1  MAGIC.10  MAGIC.94 MAGIC.534 MAGIC.692 MAGIC.693 
##       "G"       "G"       "A"       "A"       "A"       "G"</code></pre>
<p>If we want to see the genotype probability matrix for that marker:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw"><a href="../reference/getGenotypes.html">getGenotypes</a></span>(magic_geno, <span class="dt">type =</span> <span class="st">"probability"</span>)$MN1_29291)</code></pre></div>
<pre><code>##                    Bur          Can          Col           Ct          Edi
## MAGIC.1   4.009410e-08 1.598693e-02 7.209811e-01 3.180553e-07 1.807606e-03
## MAGIC.10  4.008952e-08 8.593144e-01 1.695137e-02 2.310732e-07 1.270743e-03
## MAGIC.94  8.577679e-02 1.855517e-08 1.855772e-08 2.260272e-02 1.351491e-07
## MAGIC.534 1.085882e-01 1.054246e-07 2.921143e-07 4.570506e-01 1.870980e-08
## MAGIC.692 8.578111e-02 1.855118e-08 1.855373e-08 2.260383e-02 1.351510e-07
## MAGIC.693 2.743700e-07 1.094609e-02 1.094814e-02 4.017387e-08 2.846193e-02
##                     Hi           Kn          Ler           Mt           No
## MAGIC.1   1.807741e-03 4.009246e-08 1.809008e-03 4.465485e-02 6.176561e-02
## MAGIC.10  1.265560e-03 4.016395e-08 1.265559e-03 1.695137e-02 1.695137e-02
## MAGIC.94  3.733691e-07 4.336636e-02 1.676746e-07 1.855771e-08 1.855772e-08
## MAGIC.534 1.871106e-08 1.085880e-01 1.872444e-08 2.921143e-07 2.921143e-07
## MAGIC.692 3.733832e-07 4.336851e-02 1.676779e-07 1.855373e-08 1.855373e-08
## MAGIC.693 3.018148e-02 2.534173e-07 8.000111e-01 1.094814e-02 1.094814e-02
##                     Oy           Po         Rsch           Sf          Tsu
## MAGIC.1   4.013616e-08 4.013616e-08 4.465627e-02 4.009411e-08 1.808736e-03
## MAGIC.10  4.008951e-08 4.008951e-08 1.695137e-02 4.008951e-08 1.265559e-03
## MAGIC.94  4.916271e-02 4.916271e-02 1.855771e-08 7.499277e-01 1.676687e-07
## MAGIC.534 1.085906e-01 1.085906e-01 2.921143e-07 1.085882e-01 1.872444e-08
## MAGIC.692 4.916512e-02 4.916512e-02 1.855373e-08 7.499153e-01 1.676720e-07
## MAGIC.693 5.716623e-07 5.716623e-07 1.094814e-02 2.744392e-07 4.281773e-02
##                    Wil           Ws           Wu           Zu
## MAGIC.1   4.465610e-02 2.463717e-02 1.770895e-02 1.771944e-02
## MAGIC.10  1.695137e-02 1.695137e-02 1.695685e-02 1.695262e-02
## MAGIC.94  1.855771e-08 1.855771e-08 1.857383e-08 1.859582e-08
## MAGIC.534 2.921143e-07 2.067150e-06 1.215602e-07 1.213665e-07
## MAGIC.692 1.855373e-08 1.855373e-08 1.856985e-08 1.859184e-08
## MAGIC.693 1.094814e-02 1.094656e-02 1.094622e-02 1.094622e-02</code></pre>
<p>And finally, if we want to see what the founder genotypes were in that marker:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/getFounderGenotypes.html">getFounderGenotypes</a></span>(magic_geno)$MN1_29291</code></pre></div>
<pre><code>##    accession    marker allele
## 1        Bur MN1_29291      A
## 2        Can MN1_29291      G
## 3        Col MN1_29291      G
## 4         Ct MN1_29291      A
## 5        Edi MN1_29291      G
## 6         Hi MN1_29291      G
## 7         Kn MN1_29291      A
## 8        Ler MN1_29291      G
## 9         Mt MN1_29291      G
## 10        No MN1_29291      G
## 11        Oy MN1_29291      A
## 12        Po MN1_29291      A
## 13      Rsch MN1_29291      G
## 14        Sf MN1_29291      A
## 15       Tsu MN1_29291      G
## 16       Wil MN1_29291      G
## 17        Ws MN1_29291      G
## 18        Wu MN1_29291      G
## 19        Zu MN1_29291      G</code></pre>
</div>
<div id="adding-phenotypes-to-the-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-phenotypes-to-the-analysis" class="anchor"></a>Adding phenotypes to the analysis</h2>
<p>To perform the QTL analysis, you should create a table with phenotypes. This should be a regular <code>data.frame</code>, with one of the columns containing the MAGIC line IDs.</p>
<p>In this example, we read the phenotype table that was downloaded before:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">'~/temp/magic_intro/magic_phenotype_example.txt'</span>, <span class="dt">header =</span> T)
<span class="kw">head</span>(pheno)</code></pre></div>
<pre><code>##   SUBJECT.NAME bolt.to.flower days.to.bolt days.to.germ
## 1     MAGIC.10           10.0         40.0          6.0
## 2     MAGIC.94           10.0         24.6          6.0
## 3     MAGIC.95            7.0         20.2          6.0
## 4     MAGIC.96           10.0         36.0          6.0
## 5     MAGIC.97           10.2         25.2          9.4
## 6     MAGIC.98           10.6         27.8          6.0</code></pre>
<p>To add this information to the “MagicGen” object we use the function <code><a href="../reference/addPhenotypes.html">addPhenotypes()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Pass the phenotype data.frame and specify which column contains the MAGIC line IDs</span>
magic_phen &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addPhenotypes.html">addPhenotypes</a></span>(magic_geno, pheno, <span class="st">"SUBJECT.NAME"</span>)</code></pre></div>
<p>This returns an object of class “MagicGenPhen”, which is very similar to the one from before, but with phenotype information added to it. Also, the genotypes are kept only for those individuals that had a phenotype.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">magic_phen</code></pre></div>
<pre><code>## Object of class MagicGenPhen 
## 426 MAGIC lines with a phenotype.
## There are 3 phenotypes:
##  bolt.to.flower days.to.bolt days.to.germ 
## Using genotypes for 1258 markers.</code></pre>
<p>The reason to separate these two objects, is that the “MagicGen” object only has to be created once, and different tables of phenotypes can then be added to it later on. For example, the user might consider to create this object once and then save it as an <code>.rds</code> object, to use later on in future analysis:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">saveRDS</span>(magic_geno, <span class="st">'~/temp/magic_intro/magic_genotype_probabilities.rds'</span>)</code></pre></div>
<p>In case you want to create a “MagicGenPhen” object (containing genotypes and phenotypes) in one go, you can pass the phenotype information to the <code><a href="../reference/magicFounderReconstruct.html">magicFounderReconstruct()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">magic_phen2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/magicFounderReconstruct.html">magicFounderReconstruct</a></span>(<span class="dt">snp_dir =</span> <span class="st">"~/temp/magic_intro/"</span>,
                                             <span class="dt">phenotypes =</span> pheno, <span class="dt">id =</span> <span class="st">"SUBJECT.NAME"</span>)</code></pre></div>
</div>
<div id="performing-standard-qtl-mapping" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-standard-qtl-mapping" class="anchor"></a>Performing standard QTL mapping</h2>
<p>The function <code><a href="../reference/scanQtl.html">scanQtl()</a></code> performs a basic association analysis using an F-test to compare two linear models, one with the genotype probabilities for each founder accession as explanatory variables and a null model with no explanatory variables (only an intercept).</p>
<p>For example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># QTL scan for days.to.bolt trait</span>
bolt_qtl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scanQtl.html">scanQtl</a></span>(magic_phen, <span class="st">"days.to.bolt"</span>)</code></pre></div>
<pre><code>## F-test comparing the models:
##   H1: days.to.bolt ~ GEN
##   H0: days.to.bolt ~ 1</code></pre>
<pre><code>## Elapsed time: 30 seconds</code></pre>
<p>The function supports the use of multiple cores (not supported on Windows), which can speed the computation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># QTL scan for days.to.bolt trait</span>
bolt_qtl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scanQtl.html">scanQtl</a></span>(magic_phen, <span class="st">"days.to.bolt"</span>, <span class="dt">cores =</span> <span class="dv">2</span>)
<span class="kw">head</span>(bolt_qtl)</code></pre></div>
<p>This returns a standard <code>data.frame</code> object, and it can be plotted, for example with <code>ggplot2</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load the library and change the default theme</span>
<span class="kw">library</span>(ggplot2); <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/theme_get">theme_set</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_bw</a></span>())

<span class="co"># Make the plot, with an horizontal line at 3.5, which is suggested in Kover et al. (2009)</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(bolt_qtl, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(bp/<span class="fl">1e6</span>, -<span class="kw">log10</span>(p))) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>() +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/facet_grid">facet_grid</a></span>(~<span class="st"> </span>chromosome, <span class="dt">scales =</span> <span class="st">"free_x"</span>, <span class="dt">space =</span> <span class="st">"free_x"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_abline">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="fl">3.5</span>, <span class="dt">linetype =</span> <span class="st">"dotted"</span>)</code></pre></div>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_path).</code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-15-1.png" width="672"></p>
<p>You can also get the <em>R^2</em> of linear model for the peak markers, which can give an indication of the variance explained by them. For example, the peak marker on Chr4, in the <em>Frigida</em> gene:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bolt_qtl %&gt;%<span class="st"> </span><span class="kw">arrange</span>(p) %&gt;%<span class="st"> </span><span class="kw">slice</span>(<span class="dv">1</span>)</code></pre></div>
<pre><code>## # A tibble: 1 x 10
##   marker     map     bp chromosome     f   df1   df2         p r2_h1 r2_h0
##   &lt;I(chr)&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;I(chr)&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 FRI_2343  1.13 270862 4           8.87  18.0   407  2.16e⁻²⁰ 0.250     0</code></pre>
<p>The <code>r2_h1</code> column contains the <em>R^2</em> of the linear model for the alternative hypothesis tested, suggesting that ~25% of the variance is explained by that SNP.</p>
</div>
<div id="custom-models" class="section level2">
<h2 class="hasAnchor">
<a href="#custom-models" class="anchor"></a>Custom models</h2>
<p>The <code><a href="../reference/scanQtl.html">scanQtl()</a></code> function allows to fit more complex, customised models. You can explicitly specify these models using the same sintax used for the <code>lm()</code> function in the base R package.</p>
<div id="using-covariate-traits" class="section level3">
<h3 class="hasAnchor">
<a href="#using-covariate-traits" class="anchor"></a>Using covariate traits</h3>
<p>For example, the traits <em>days.to.bolt</em> and <em>bolt.to.flower</em> are correlated with each other:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/qplot">qplot</a></span>(days.to.bolt, bolt.to.flower, <span class="dt">data =</span> pheno)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<p>Indeed, the QTL scan for the two traits picks similar QTL:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">flower_qtl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scanQtl.html">scanQtl</a></span>(magic_phen, <span class="st">"bolt.to.flower"</span>, <span class="dt">cores =</span> <span class="dv">2</span>)</code></pre></div>
<pre><code>## F-test comparing the models:
##   H1: bolt.to.flower ~ GEN
##   H0: bolt.to.flower ~ 1</code></pre>
<pre><code>## Elapsed time: 22 seconds</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(flower_qtl, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(bp/<span class="fl">1e6</span>, -<span class="kw">log10</span>(p))) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>() +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/facet_grid">facet_grid</a></span>(~<span class="st"> </span>chromosome, <span class="dt">scales =</span> <span class="st">"free_x"</span>, <span class="dt">space =</span> <span class="st">"free_x"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_abline">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="fl">3.5</span>, <span class="dt">linetype =</span> <span class="st">"dotted"</span>)</code></pre></div>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_path).</code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<p>But let’s say we wanted to find QTL that explain some of the residual variation that might exist when the correlation between the two traits is taken into account. For this, we could use one of them as a covariate in the model, for example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Run QTL scan for "days.to.bolt" but adding "bolt.to.flower" as a covariate</span>
bolt_cov_qtl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scanQtl.html">scanQtl</a></span>(magic_phen, <span class="st">"days.to.bolt"</span>,
                             <span class="dt">covariates =</span> <span class="st">"bolt.to.flower"</span>)</code></pre></div>
<pre><code>## F-test comparing the models:
##   H1:days.to.bolt ~ GEN + bolt.to.flower
##   H0:days.to.bolt ~ bolt.to.flower</code></pre>
<pre><code>## Elapsed time: 37 seconds</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Make plot</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(bolt_cov_qtl, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(bp/<span class="fl">1e6</span>, -<span class="kw">log10</span>(p))) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">colour =</span> <span class="st">"covariate"</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/facet_grid">facet_grid</a></span>(~<span class="st"> </span>chromosome, <span class="dt">scales =</span> <span class="st">"free_x"</span>, <span class="dt">space =</span> <span class="st">"free_x"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_abline">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="fl">3.5</span>, <span class="dt">linetype =</span> <span class="st">"dotted"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>(<span class="dt">data =</span> bolt_qtl, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">colour =</span> <span class="st">"single trait"</span>))</code></pre></div>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_path).

## Warning: Removed 2 rows containing missing values (geom_path).</code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-19-1.png" width="672"></p>
<p>We can see that the QTL on Chr4 is still significant, even when the correlation between the two traits is taken into account, whereas those on Chr1 and Chr5 are less convincing.</p>
</div>
<div id="using-snp-covariates" class="section level3">
<h3 class="hasAnchor">
<a href="#using-snp-covariates" class="anchor"></a>Using SNP covariates</h3>
<p>We might also want to explore whereas different SNPs explain common variance in our trait.</p>
<p>To explore this hypothesis, we can also use other markers as covariates in the QTL model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Run QTL scan conditional on the peak marker on Chr5</span>
peak5 &lt;-<span class="st"> </span>bolt_qtl %&gt;%<span class="st"> </span><span class="kw">filter</span>(chromosome ==<span class="st"> </span><span class="dv">5</span>) %&gt;%<span class="st"> </span><span class="kw">arrange</span>(p) %&gt;%<span class="st"> </span><span class="kw">slice</span>(<span class="dv">1</span>) %&gt;%<span class="st"> </span>.$marker
bolt_snp_qtl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scanQtl.html">scanQtl</a></span>(magic_phen, <span class="st">"days.to.bolt"</span>,
                             <span class="dt">marker_cov =</span> peak5)</code></pre></div>
<pre><code>## F-test comparing the models:
##   H1:days.to.bolt ~ GEN + MN5_3177504
##   H0:days.to.bolt ~ MN5_3177504</code></pre>
<pre><code>## Elapsed time: 33 seconds</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(bolt_snp_qtl, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(bp/<span class="fl">1e6</span>, -<span class="kw">log10</span>(p))) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">colour =</span> <span class="st">"FRI_2343"</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/facet_grid">facet_grid</a></span>(~<span class="st"> </span>chromosome, <span class="dt">scales =</span> <span class="st">"free_x"</span>, <span class="dt">space =</span> <span class="st">"free_x"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_abline">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="fl">3.5</span>, <span class="dt">linetype =</span> <span class="st">"dotted"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>(<span class="dt">data =</span> bolt_qtl, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">colour =</span> <span class="st">"single trait"</span>))</code></pre></div>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_path).

## Warning: Removed 2 rows containing missing values (geom_path).</code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-20-1.png" width="672"></p>
<p>Since the two other QTL on Chr1 and Chr4 remain after taking into account the contribution of the QTL on Chr5, it suggests that they explain extra phenotypic variance not captured by that QTL.</p>
</div>
<div id="custom-models-1" class="section level3">
<h3 class="hasAnchor">
<a href="#custom-models-1" class="anchor"></a>Custom models</h3>
<p>(sorry, proper example to be added…)</p>
<p>But essentially you can use the same syntax used in the <code>lm()</code> function in the base R package, to define the null and alternative hypothesis of the models being compared (these are the options <code>h1</code> and <code>h0</code> in the <code><a href="../reference/scanQtl.html">scanQtl()</a></code> function). The genotype term is specified as GEN. The message printed when running the above <code><a href="../reference/scanQtl.html">scanQtl()</a></code> calls gives examples of how the models are specified internally.</p>
<p>This allows for example running models using interaction terms, etc.</p>
</div>
<div id="even-more-customised-models" class="section level3">
<h3 class="hasAnchor">
<a href="#even-more-customised-models" class="anchor"></a>Even more customised models</h3>
<p>If you feel limited by the capability of the <code><a href="../reference/scanQtl.html">scanQtl()</a></code> function, you can make your own implementation, by accessing the data in the <em>MagicGenPhen</em> object. For example, here is a “manual” implementation of the basic QTL scan:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">phenotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPhenotypes.html">getPhenotypes</a></span>(magic_phen)
genotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGenotypes.html">getGenotypes</a></span>(magic_phen) <span class="co"># by default returns genotype probabilities</span>
markers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getMarkers.html">getMarkers</a></span>(magic_phen)

<span class="co"># Run the association test for each marker and retrieve the p-value</span>
pvals &lt;-<span class="st"> </span><span class="kw">sapply</span>(genotypes, function(GEN, PHEN){
  <span class="co"># Fit the alternative</span>
  fit1 &lt;-<span class="st"> </span><span class="kw">lm</span>(PHEN ~<span class="st"> </span>GEN)
  
  <span class="co"># Fit the null</span>
  fit0 &lt;-<span class="st"> </span><span class="kw">lm</span>(PHEN ~<span class="st"> </span><span class="dv">1</span>)
  
  <span class="co"># Perform the F-test and retrieve the P-value</span>
  <span class="kw">anova</span>(fit0, fit1)$<span class="st">`</span><span class="dt">Pr(&gt;F)</span><span class="st">`</span>[<span class="dv">2</span>]
  
}, <span class="dt">PHEN =</span> phenotypes$days.to.bolt)

<span class="co"># Merge with marker positions</span>
pvals &lt;-<span class="st"> </span><span class="kw">cbind</span>(markers, <span class="dt">p =</span> pvals)

<span class="co"># Plot</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(pvals, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(bp/<span class="fl">1e6</span>, -<span class="kw">log10</span>(p))) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">colour =</span> <span class="st">"manual"</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/facet_grid">facet_grid</a></span>(~<span class="st"> </span>chromosome, <span class="dt">scales =</span> <span class="st">"free_x"</span>, <span class="dt">space =</span> <span class="st">"free_x"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_abline">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="fl">3.5</span>, <span class="dt">linetype =</span> <span class="st">"dotted"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>(<span class="dt">data =</span> bolt_qtl, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">colour =</span> <span class="st">"scanQtl"</span>), <span class="dt">linetype =</span> <span class="st">"dashed"</span>)</code></pre></div>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_path).

## Warning: Removed 2 rows containing missing values (geom_path).</code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
</div>
</div>
<div id="estimating-qtl-effect-sizes" class="section level1">
<h1 class="hasAnchor">
<a href="#estimating-qtl-effect-sizes" class="anchor"></a>Estimating QTL effect sizes</h1>
<p>Once a significant QTL peak is identified, we might be interested in what the predicted effect of each allele is in our phenotype. Because the MAGIC line genotypes in each marker are assigned to each parent as a probability, this has to be done by a probabilistic “phenotype imputation”. Details of the method can be found in <a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1000551">Kover P .. Mott R (2009)</a>, but essentially, we can use the function <code><a href="../reference/estimateFounderEffect.html">estimateFounderEffect()</a></code> to achieve this goal:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Predicted effect of QTL on chr5 identified previously</span>
peak5_effect &lt;-<span class="st"> </span><span class="kw"><a href="../reference/estimateFounderEffect.html">estimateFounderEffect</a></span>(magic_phen, <span class="st">"days.to.bolt"</span>, peak5)</code></pre></div>
<pre><code>## Warning: Column `accession` joining factor and character vector, coercing
## into character vector</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(peak5_effect)</code></pre></div>
<pre><code>## # A tibble: 6 x 7
##   accession effect_mean effect_up effect_lo     n marker      allele
##   &lt;fct&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt; 
## 1 Bur           -0.213    -0.196     -0.233 37.0  MN5_3177504 T     
## 2 Can            0.748     0.814      0.550  7.00 MN5_3177504 T     
## 3 Col            0.517     0.555      0.439 15.0  MN5_3177504 T     
## 4 Ct            -0.119     0.0366    -0.232 16.0  MN5_3177504 C     
## 5 Edi            0.635     0.895      0.373  9.00 MN5_3177504 C     
## 6 Hi            -0.0874   -0.0254    -0.130 37.0  MN5_3177504 C</code></pre>
<p>This returns a <code>data.frame</code> with the predicted contribution of each accession to the phenotype. Because this is probabilistic in nature, the method relies on repeating this estimate several times (500 by default). By default, the <code><a href="../reference/estimateFounderEffect.html">estimateFounderEffect()</a></code> function returns a summary of these probabilistic estimates with a mean and lower and upper “confidence interval” for the estimate.</p>
<p>By default, it also standardizes the trait as z-scores (i.e. subtracts the mean trait value observed in the sampled individuals from the estimate and divides by the observed standard deviation - in other words, the values can be read as “standard deviations away from the mean”).</p>
<p>The result can be plotted like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(peak5_effect, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(accession, effect_mean)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_linerange">geom_pointrange</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">ymin =</span> effect_lo, <span class="dt">ymax =</span> effect_up))</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-23-1.png" width="672"></p>
<p>We can see from this example that, at this locus:</p>
<ul>
<li>The <em>Zu</em> allele should result in later-flowering ~1.5 SD above the mean</li>
<li>The <em>Ler</em> allele should result in early-flowering ~0.7 SD below the mean</li>
</ul>
<p>The errorbars also show that some estimates are better than others.</p>
<p>If you prefer to have the effect size in the units of your trait, then you could achieve a similar result by:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/estimateFounderEffect.html">estimateFounderEffect</a></span>(magic_phen, <span class="st">"days.to.bolt"</span>, peak5, <span class="dt">standardised =</span> <span class="ot">FALSE</span>) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(accession, effect_mean)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_linerange">geom_pointrange</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">ymin =</span> effect_lo, <span class="dt">ymax =</span> effect_up))</code></pre></div>
<pre><code>## Warning: Column `accession` joining factor and character vector, coercing
## into character vector</code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-24-1.png" width="672"></p>
<p>The plot is essentially the same, but now the scale can be read in the original units of “days.to.flower”, so between ~20 days for <em>Ler</em> and ~32 days for <em>Zu</em>.</p>
<p>Another trick that can be used is to add what the SNP allele is for each of the founder accessions. This is also available on the output from <code><a href="../reference/estimateFounderEffect.html">estimateFounderEffect()</a></code> function, and can be visualised like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(peak5_effect, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(accession, effect_mean)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_linerange">geom_pointrange</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">ymin =</span> effect_lo, <span class="dt">ymax =</span> effect_up)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_text">geom_text</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">y =</span> -<span class="dv">1</span>, <span class="dt">label =</span> allele, <span class="dt">colour =</span> allele)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_manual">scale_color_manual</a></span>(<span class="dt">values =</span> <span class="kw">c</span>(<span class="st">"black"</span>, <span class="st">"red3"</span>)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/theme">theme</a></span>(<span class="dt">legend.position =</span> <span class="st">"none"</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-25-1.png" width="672"></p>
<p>(note that you could also get the founder genotypes with <code>getFounderGenotypes(magic_phen)[[peak5]]</code>)</p>
<p>This nicely shows that the effect of the SNP allele is not that well correlated with the predicted effect of the founder allele. For example, accessions <em>Kn</em> and <em>Zu</em> have very different predicted flowering time contributions, but they both have the same SNP genotype at this marker.</p>
<p>Another way to consider this, is to group the MAGIC lines by their SNP allele and plot the phenotype distributions for this. This can be achieved by another function called <code><a href="../reference/inferMagicFounder.html">inferMagicFounder()</a></code>, which assigns each MAGIC line to a single founder, based on some probability threshold (0.5 by default).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">peak5_inferred &lt;-<span class="st"> </span><span class="kw"><a href="../reference/inferMagicFounder.html">inferMagicFounder</a></span>(magic_phen, peak5)</code></pre></div>
<pre><code>## 423 MAGIC lines were attributed to exactly one founder.</code></pre>
<pre><code>## 3 MAGIC lines were not attributed a founder.</code></pre>
<pre><code>## 0 MAGIC lines were attributed more than one founder.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(peak5_inferred)</code></pre></div>
<pre><code>##       magic founder bolt.to.flower days.to.bolt days.to.germ allele
## 1  MAGIC.10      Zu           10.0         40.0          6.0      C
## 2 MAGIC.101      Mt           11.2         21.4          6.2      C
## 3 MAGIC.102      Kn            7.0         19.4          6.0      C
## 4 MAGIC.103     Col           10.0         23.6          6.0      T
## 5 MAGIC.104     Wil           10.4         30.4          6.0      T
## 6 MAGIC.105      Sf           13.0         39.4          7.0      T</code></pre>
<p>This table thus assigns each MAGIC line to <strong>one</strong> founder only, if the probability of being that founder is greater than 50%. The threshold can be increased, and this usually results in more MAGIC lines not having an assigned parental genotype.</p>
<p>This table also returns the SNP genotypes for that marker. Therefore, we can produce the phenotype distribution grouped by SNP allele:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(peak5_inferred, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(allele, days.to.bolt)) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_boxplot">geom_boxplot</a></span>()</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-27-1.png" width="672"></p>
<p>From this we can see that the average (or median) phenotype of the two alleles (C and T) are not that different, consistent with a lack of correlation in the predicted accession effects above.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#downloading-the-data">Downloading the data</a></li>
      <li><a href="#genotype-probability-matrix">Genotype probability matrix</a></li>
      <li><a href="#adding-phenotypes-to-the-analysis">Adding phenotypes to the analysis</a></li>
      <li><a href="#performing-standard-qtl-mapping">Performing standard QTL mapping</a></li>
      <li><a href="#custom-models">Custom models</a></li>
      <li><a href="#estimating-qtl-effect-sizes">Estimating QTL effect sizes</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hugo Tavares.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
