<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Retrieving genes in QTL • MagicHelpR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">MagicHelpR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Introduction to MagicHelpR</a>
    </li>
    <li>
      <a href="../articles/retrieving_genes_in_qtl.html">Retrieving genes in QTL</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Retrieving genes in QTL</h1>
                        <h4 class="author">Hugo Tavares</h4>
            
            <h4 class="date">2 August 2017</h4>
          </div>

    
    
<div class="contents">
<p>This vignette shows an example of how to retrieve gene lists around QTL of interest. It’s still in progress and simply has some ideas of how this might be done.</p>
<p>At the moment there’s no method implemented in <code>MagicHelpR</code> to define what a QTL peak is (or its confidence interval). In the original MAGIC paper they have a method to define confidence intervals for the QTL, but I could not get my head around it yet.</p>
<p>More naively, one could define a QTL peak as a region within a certain Kb distance from the top SNP. This is the starting point in this vignette (perhaps in the future this will be improved).</p>
<div id="prepare-data" class="section level1">
<h1 class="hasAnchor">
<a href="#prepare-data" class="anchor"></a>Prepare data</h1>
<p>I use the same data shown in the <a href="https://tavareshugo.github.io/MagicHelpR//articles/introduction.html">introduction vignette</a>.</p>
<p>I will use flowering time as an example.</p>
<p>The code to reproduce the analysis shown in that vignette is here for convenience:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(MagicHelpR)</code></pre></div>
<pre><code>## Loading required package: dplyr</code></pre>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre><code>## Loading required package: tidyr</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Download genotype and phenotype example data</span>
<span class="co">#downloadArabMagic('~/temp/magic_intro/', tidy = TRUE, example_data = TRUE)</span>

<span class="co"># Read phenotype data</span>
pheno &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">'~/temp/magic_intro/magic_phenotype_example.txt'</span>, <span class="dt">header =</span> T)

<span class="co"># Create the MagicGenPhen object containing genotypes and phenotype</span>
magic_phen &lt;-<span class="st"> </span><span class="kw"><a href="../reference/magicFounderReconstruct.html">magicFounderReconstruct</a></span>(<span class="dt">snp_dir =</span> <span class="st">"~/temp/magic_intro/"</span>,
                                                                            <span class="dt">phenotypes =</span> pheno, <span class="dt">id =</span> <span class="st">"SUBJECT.NAME"</span>)</code></pre></div>
<pre><code>## mindist: 1e-05
## datafile /home/hugo/temp/magic_intro//all_chr.MAGIC.happy.data allelesfile /home/hugo/temp/magic_intro//all_chr.MAGIC.happy.alleles gen 7
## genotype phase: unknown
## Reading phenotype and genotype data from ped file /home/hugo/temp/magic_intro//all_chr.MAGIC.happy.data
## Number of individuals: 703  
## Number of markers:     1259 
## Number of strains:     19   
## Use Parents:           no
## Number of subjects with two parents: 0    
## Number of parents in nuclear families: 0    
## null model mean nan var nan
## assuming haploid(inbred) genotypes
## datafile /home/hugo/temp/magic_intro//all_chr.MAGIC.happy.data allelesfile /home/hugo/temp/magic_intro//all_chr.MAGIC.happy.alleles gen 7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># QTL scan for days.to.bolt trait</span>
bolt_qtl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scanQtl.html">scanQtl</a></span>(magic_phen, <span class="st">"days.to.bolt"</span>)</code></pre></div>
<pre><code>## F-test comparing the models:
##   H1: days.to.bolt ~ GEN
##   H0: days.to.bolt ~ 1</code></pre>
<pre><code>## Elapsed time: 31 seconds</code></pre>
<p>This is our QTL plot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load the library and change the default theme</span>
<span class="kw">library</span>(ggplot2); <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/theme_get">theme_set</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_bw</a></span>())

<span class="co"># Make the plot, with an horizontal line at 3.5, which is suggested in Kover et al. (2009)</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(bolt_qtl, <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(bp/<span class="fl">1e6</span>, -<span class="kw">log10</span>(p))) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_line</a></span>() +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/facet_grid">facet_grid</a></span>(~<span class="st"> </span>chromosome, <span class="dt">scales =</span> <span class="st">"free_x"</span>, <span class="dt">space =</span> <span class="st">"free_x"</span>) +
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_abline">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="fl">3.5</span>, <span class="dt">linetype =</span> <span class="st">"dotted"</span>)</code></pre></div>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_path).</code></pre>
<p><img src="retrieving_genes_in_qtl_files/figure-html/unnamed-chunk-2-1.png" width="672"></p>
</div>
<div id="retrieving-genes-around-a-qtl" class="section level1">
<h1 class="hasAnchor">
<a href="#retrieving-genes-around-a-qtl" class="anchor"></a>Retrieving genes around a QTL</h1>
<p>Let’s say we were interested in which genes are around the peak on Chr4.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Filter the data to retain only the peak marker on Chromosome 4</span>
chr4_peak &lt;-<span class="st"> </span>bolt_qtl %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">filter</span>(chromosome ==<span class="st"> </span><span class="dv">4</span>) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">arrange</span>(p) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">slice</span>(<span class="dv">1</span>) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">select</span>(marker, bp, chromosome)

chr4_peak</code></pre></div>
<pre><code>## # A tibble: 1 x 3
##   marker       bp chromosome
##   &lt;I(chr)&gt;  &lt;dbl&gt; &lt;I(chr)&gt;  
## 1 FRI_2343 270862 4</code></pre>
<p>As I said above, at the moment I don’t know what the best way is to define where a peak starts and ends.</p>
<p>However, as a starting point, let’s consider all genes within 50Kb of the top SNP on Chr4.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Add positions for window start and end</span>
chr4_peak &lt;-<span class="st"> </span>chr4_peak %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">window_start =</span> bp -<span class="st"> </span><span class="fl">50e3</span>,
                 <span class="dt">window_end =</span> bp +<span class="st"> </span><span class="fl">50e3</span>)

chr4_peak</code></pre></div>
<pre><code>## # A tibble: 1 x 5
##   marker       bp chromosome window_start window_end
##   &lt;I(chr)&gt;  &lt;dbl&gt; &lt;I(chr)&gt;          &lt;dbl&gt;      &lt;dbl&gt;
## 1 FRI_2343 270862 4                220862     320862</code></pre>
<p>I will use the package <a href="https://bioconductor.org/packages/biomaRt/"><code>biomaRt</code></a> to retrieve gene information from <a href="www.ensembl.org/biomart">BioMart</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"biomaRt"</span>)

<span class="co"># List of BioMart plant databases</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/listMarts">listMarts</a></span>(<span class="dt">host=</span><span class="st">"plants.ensembl.org"</span>)</code></pre></div>
<pre><code>##             biomart                      version
## 1       plants_mart      Ensembl Plants Genes 38
## 2 plants_variations Ensembl Plants Variations 38</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Specify the plants_mart</span>
m &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"plants_mart"</span>, <span class="dt">host=</span><span class="st">"plants.ensembl.org"</span>)

<span class="co"># Check the Arabidopsis dataset name</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/listDatasets">listDatasets</a></span>(m) %&gt;%<span class="st"> </span><span class="kw">filter</span>(<span class="kw">grepl</span>(<span class="st">"Arabidopsis"</span>, description))</code></pre></div>
<pre><code>##             dataset                         description version
## 1 athaliana_eg_gene Arabidopsis thaliana genes (TAIR10)  TAIR10
## 2   alyrata_eg_gene    Arabidopsis lyrata genes (v.1.0)   v.1.0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Define dataset</span>
m &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"plants_mart"</span>, <span class="dt">host=</span><span class="st">"plants.ensembl.org"</span>, <span class="dt">dataset=</span><span class="st">"athaliana_eg_gene"</span>)</code></pre></div>
<p>The function <code>getBM</code> can then be used to retrieve genes from the BioMart database.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Download gene list with certain attributes</span>
## attribute list can be obtained with listAttributes(m)
biomart_genes &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM</a></span>(<span class="dt">attributes=</span><span class="kw">c</span>(<span class="st">"tair_locus"</span>, <span class="st">"tair_symbol"</span>, <span class="st">"external_gene_name"</span>,
                                    <span class="st">"chromosome_name"</span>, <span class="st">"start_position"</span>, <span class="st">"end_position"</span>,
                                    <span class="st">"description"</span>),
                       <span class="dt">mart=</span>m) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(tair_locus !=<span class="st"> ""</span>)  <span class="co"># some attributes have no "tair_locus" ID!</span>

<span class="kw">head</span>(biomart_genes)</code></pre></div>
<pre><code>##   tair_locus tair_symbol external_gene_name chromosome_name start_position
## 1  AT5G16970      AT-AER                AER               5        5575973
## 2  AT4G32100                                              4       15511757
## 3  AT2G43120                                              2       17927181
## 4  AT1G30814                                              1       10942648
## 5  AT3G18710       PUB29              PUB29               3        6434082
## 6  AT4G25880       APUM6              APUM6               4       13154708
##   end_position
## 1      5578086
## 2     15512218
## 3     17929138
## 4     10944727
## 5      6436046
## 6     13159382
##                                                                                                                                                                                                                                 description
## 1                                                                                                                                               NADPH-dependent oxidoreductase 2-alkenal reductase [Source:UniProtKB/Swiss-Prot;Acc:Q39172]
## 2                                                                                                                                              Beta-1,3-N-Acetylglucosaminyltransferase family protein [Source:UniProtKB/TrEMBL;Acc:F4JTI5]
## 3                                                                                                                                                                          RmlC-like cupins superfamily protein [Source:TAIR;Acc:AT2G43120]
## 4 unknown protein; Has 35333 Blast hits to 34131 proteins in 2444 species: Archae - 798; Bacteria - 22429; Metazoa - 974; Fungi - 991; Plants - 531; Viruses - 0; Other Eukaryotes - 9610 (source: NCBI BLink). [Source:TAIR;Acc:AT1G30814]
## 5                                                                                                                                                               RING-type E3 ubiquitin transferase [Source:UniProtKB/TrEMBL;Acc:A0A178VJJ8]
## 6                                                                                                                                                                                                     pumilio 6 [Source:TAIR;Acc:AT4G25880]</code></pre>
<p>This table should contain all Arabidopsis genes from BioMart.</p>
<p>We can then filter it based on our peak QTL:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chr4_genes &lt;-<span class="st"> </span>biomart_genes %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(chromosome_name ==<span class="st"> </span>chr4_peak$chromosome &amp;
<span class="st">         </span>start_position &gt;<span class="st"> </span>chr4_peak$window_start &amp;<span class="st"> </span>
<span class="st">           </span>end_position &lt;<span class="st"> </span>chr4_peak$window_end)</code></pre></div>
<p>There’s now 29 in this list. We can further arrange them by their closeness to our top SNP:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get the middle position of each gene and get how far it is from the top SNP</span>
<span class="co"># arrange by distance and get the first few rows of the table</span>
chr4_genes %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">mid_position =</span> <span class="kw">round</span>((start_position +<span class="st"> </span>end_position)/<span class="dv">2</span>),
           <span class="dt">dist_snp =</span> <span class="kw">abs</span>(mid_position -<span class="st"> </span>chr4_peak$bp)) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">arrange</span>(dist_snp) %&gt;%
<span class="st">    </span><span class="kw">head</span>()</code></pre></div>
<pre><code>##   tair_locus tair_symbol external_gene_name chromosome_name start_position
## 1  AT4G00650         FRI                FRI               4         268901
## 2  AT4G00651                                              4         271487
## 3  AT4G00660         RH8                RH8               4         274252
## 4  AT4G00630        KEA2               KEA2               4         261246
## 5  AT4G00670                                              4         278316
## 6  AT4G00680        ADF8               ADF8               4         279596
##   end_position
## 1       271503
## 2       271879
## 3       278737
## 4       268097
## 5       279197
## 6       281072
##                                                                                                                                                                                                                                                                                                                                                                                                        description
## 1                                                                                                                                                                                                                                                                                                                                                Inactive protein FRIGIDA [Source:UniProtKB/Swiss-Prot;Acc:Q67Z93]
## 2 FUNCTIONS IN: molecular_function unknown; INVOLVED IN: biological_process unknown; LOCATED IN: endomembrane system; BEST Arabidopsis thaliana protein match is: Fe superoxide dismutase 2 (TAIR:AT5G51100.1); Has 34 Blast hits to 34 proteins in 11 spec /.../rchae - 0; Bacteria - 0; Metazoa - 0; Fungi - 0; Plants - 34; Viruses - 0; Other Eukaryotes - 0 (source: NCBI BLink). [Source:TAIR;Acc:AT4G00651]
## 3                                                                                                                                                                                                                                                                                                                                   DEAD-box ATP-dependent RNA helicase 8 [Source:UniProtKB/Swiss-Prot;Acc:Q8RXK6]
## 4                                                                                                                                                                                                                                                                                                                                                      K+ efflux antiporter 2 [Source:UniProtKB/TrEMBL;Acc:F4JHE9]
## 5                                                                                                                                                                                                                                                                                                                                                                   At4g00670 [Source:UniProtKB/TrEMBL;Acc:Q6IDB4]
## 6                                                                                                                                                                                                                                                                                                                                                        actin depolymerizing factor 8 [Source:TAIR;Acc:AT4G00680]
##   mid_position dist_snp
## 1       270202      660
## 2       271683      821
## 3       276494     5632
## 4       264672     6190
## 5       278756     7894
## 6       280334     9472</code></pre>
<p>In this case, the gene <code>FRI</code> is the closest to our top SNP, and it is likely to be a good candidate (as it is involved in flowering).</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#prepare-data">Prepare data</a></li>
      <li><a href="#retrieving-genes-around-a-qtl">Retrieving genes around a QTL</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hugo Tavares.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
